content/blog/2022/2022-08-02-performance-with-completablefutures.md

```diff
@@ -0,0 +1,324 @@
+
+## What Is A `Future`?
+
+`Future` is a Java interface that was introduced in Java 5 to represent a value that will be available in the future.
+The advantages of using a `Future` are enormous because we could do some very intensive computation asynchronously
+without blocking the current thread that in the meantime can do some other useful job.
+
+We can think of it as going to the restaurant. During the time that the chef is preparing our dinner, we can do other
+things, like talking to friends or drinking a glass of wine and once the chef has finished the preparation, we can
+finally eat. Another advantage is that using the `Future` interface is much more developer-friendly than working
+directly with threads.
+
+## Limitations Of `Future`
+
+- We cannot complete manually the operation by providing a value  
+  Let's imagine that in our case, the categorization service that we are using to get the category for our transactions,
+  goes down, we would like to complete the  `Future` manually by providing a default value. This is not possible using
+  the `Future` interface.
+- We cannot combine one or more asynchronous operations  
+  Using the same example of categorization service, we would like to call another API passing the value of the category
+  returned by our categorization service, creating a pipeline. But this scenario is not possible using the `Future`.
+- There is no way to react to the completion of the asynchronous operation without blocking the thread by
+  invoking [get](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#get())
+  method of the `Future` interface
+
+## `CompletableFuture` Vs. `Future`
+
+All the above limitations are made possible by the new class `CompletableFuture` that was introduced with  
+version 8 of Java. `CompletableFuture` implements both the `Future`
+and the [`CompletionStage`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html)
+interfaces. Let's see now how we can manage each of the limitations discussed above using `CompletableFuture` class.
+
+- We can complete manually the operation by providing a value
+
+ ```java
+class Demo {
+
+    public static void main(String[] args) {
+        CompletableFuture<String> stringCompletableFuture = new CompletableFuture<>();
+        stringCompletableFuture.complete("completed");
+        System.out.println("Is the stringCompletableFuture done ? " + stringCompletableFuture.isDone());
+    }
+}
+```
+
+Here we are creating a `CompletableFuture` of type `String` and we are completing it by using the method `complete()`
+passing a value. In the end, we are testing if really `stringCompletableFuture` has a value by using the
+method `isDone()` which returns `true` if completed in any fashion: normally, exceptionally, or via cancellation. The
+output of the `main()` method is:
+
+`Is the stringCompletableFuture done ? true`
+
+- We can combine one or more asynchronous operations
```

Comment:
Make these real section headers instead of list items (see my comment above for suggestions for header titles.

Reply:
done